<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>ULM-vignette • ULM</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ULM-vignette">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ULM</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/ULM-vignette.html">ULM-vignette</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Sidex71/ULM/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>ULM-vignette</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Sidex71/ULM/blob/HEAD/vignettes/ULM-vignette.Rmd" class="external-link"><code>vignettes/ULM-vignette.Rmd</code></a></small>
      <div class="d-none name"><code>ULM-vignette.Rmd</code></div>
    </div>

    
    
<p>Introduction ULM is a package to reconstruct physical cell-cell
interaction networks from conventional scRNAseq datasets using
signature-based approach. Basically, ULM utilises univariate linear
models to identify multiplets (mostly doublets) which potentially
represent undissociated cell fractions that are physically connected
cell neighbors in tissue.</p>
<p><img src="../../../../scripts/package_figure/algorithm.jpeg" style="width:100.0%"></p>
<p>We first load the ULM package</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Sidex71/ULM" class="external-link">ULM</a></span><span class="op">)</span></span></code></pre></div>
<p>In this vignette, we first test ULM on a scRNAseq dataset of small
intestinal tissue (Andrews et al., 2021, pubmed: GSE175664). The
preprocessed scRNAseq can be loaded by running this code:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"int_singData"</span><span class="op">)</span></span>
<span><span class="va">int_singData</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 15610 features across 5279 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (15610 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span></code></pre></div>
<p>Let us visualise the metadata to see the available cell
annotations</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">int_singData</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#&gt;                     orig.ident nCount_RNA nFeature_RNA          Cell_Type</span></span>
<span><span class="co">#&gt; AAACGAAAGAGGTCGT SeuratProject       6540         2190   Progenitor early</span></span>
<span><span class="co">#&gt; AAACGAAAGCCTCCAG SeuratProject      11193         3062  Progenitor late-1</span></span>
<span><span class="co">#&gt; AAACGAAAGTTTGGCT SeuratProject      12139         3234   Progenitor early</span></span>
<span><span class="co">#&gt; AAACGAACACAAGCTT SeuratProject      12188         3445 Transit amplifying</span></span>
<span><span class="co">#&gt; AAACGAAGTCGCTTGG SeuratProject       2493         1112  Progenitor late-2</span></span>
<span><span class="co">#&gt; AAACGAAGTCTTACTT SeuratProject       7547         2686             Goblet</span></span>
<span><span class="co">#&gt;                                       Source Subject_strain          Tissue</span></span>
<span><span class="co">#&gt; AAACGAAAGAGGTCGT Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAAAGCCTCCAG Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAAAGTTTGGCT Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAACACAAGCTT Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAAGTCGCTTGG Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAAGTCTTACTT Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt;                  Pass_QC RNA_snn_res.0.3 seurat_clusters</span></span>
<span><span class="co">#&gt; AAACGAAAGAGGTCGT    TRUE               2               2</span></span>
<span><span class="co">#&gt; AAACGAAAGCCTCCAG    TRUE               3               3</span></span>
<span><span class="co">#&gt; AAACGAAAGTTTGGCT    TRUE               0               0</span></span>
<span><span class="co">#&gt; AAACGAACACAAGCTT    TRUE               0               0</span></span>
<span><span class="co">#&gt; AAACGAAGTCGCTTGG    TRUE               6               6</span></span>
<span><span class="co">#&gt; AAACGAAGTCTTACTT    TRUE               1               1</span></span></code></pre></div>
<p>As we can see, there is a ‘Cell_Type’ column that contains cell
annotations. Let us visualize the annotated Seurat clusters</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; 'SeuratObject' was built under R 4.4.1 but the current version is</span></span>
<span><span class="co">#&gt; 4.4.3; it is recomended that you reinstall 'SeuratObject' as the ABI</span></span>
<span><span class="co">#&gt; for R may have changed</span></span>
<span><span class="co">#&gt; 'SeuratObject' was built with package 'Matrix' 1.6.5 but the current</span></span>
<span><span class="co">#&gt; version is 1.7.0; it is recomended that you reinstall 'SeuratObject' as</span></span>
<span><span class="co">#&gt; the ABI for 'Matrix' may have changed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, t</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">int_singData</span>, reduction<span class="op">=</span><span class="st">"umap"</span>, group.by<span class="op">=</span><span class="st">"Cell_Type"</span>, label<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="ULM-vignette_files/figure-html/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Now let us use the ULM pipeline to identify multiplets present in
this data and infer physical interaction network.</p>
<p>The first step in the pipeline is to generate cell type-specific gene
signatures. This will generate a default of 100 genes that best mark
each annotated cell type cluster in the data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">101324</span><span class="op">)</span></span>
<span><span class="va">int_sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetSignature.html">GetSignature</a></span><span class="op">(</span><span class="va">int_singData</span>, ident_col <span class="op">=</span> <span class="va">int_singData</span><span class="op">$</span><span class="va">Cell_Type</span>, n <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">#&gt; using the specified seurat ident to generate signatures</span></span>
<span><span class="co">#&gt; Calculating cluster Progenitor early</span></span>
<span><span class="co">#&gt; Calculating cluster Progenitor late-1</span></span>
<span><span class="co">#&gt; Calculating cluster Transit amplifying</span></span>
<span><span class="co">#&gt; Calculating cluster Progenitor late-2</span></span>
<span><span class="co">#&gt; Calculating cluster Goblet</span></span>
<span><span class="co">#&gt; Calculating cluster Stem</span></span>
<span><span class="co">#&gt; Calculating cluster Enterocyte</span></span>
<span><span class="co">#&gt; Calculating cluster Paneth</span></span>
<span><span class="co">#&gt; Calculating cluster Enteroendocrine</span></span>
<span><span class="co">#&gt; Calculating cluster Tuft</span></span></code></pre></div>
<p>Let us visualize what the signature table looks like</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">int_sig</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10 × 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   source [1]</span></span></span>
<span><span class="co">#&gt;    source           target          mor</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Progenitor early Ccna1             1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Progenitor early Hist1h2br         1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Progenitor early BC051077          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Progenitor early Gm10501           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Progenitor early C330021F23Rik     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Progenitor early Cdkn3             1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Progenitor early Cdc25c            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Progenitor early Ccnb2             1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Progenitor early Cdc20             1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Progenitor early Lipg              1</span></span></code></pre></div>
<p>The signature table has 3 columns, the source column contains the
cell type, the target column contains the genes and the mor column
contains the weights.</p>
<p>Next, we score each cell in the scRNseq data for cell signatures by
fitting univariate linear models. We will score the signatures on each
barcode (cell) in the normalised gene count matrix (found in the slot =
data). To use the raw count, set slot = ‘count’</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetCellScores.html">GetCellScores</a></span><span class="op">(</span>seurat_obj <span class="op">=</span> <span class="va">int_singData</span>, signatures <span class="op">=</span> <span class="va">int_sig</span>, assay <span class="op">=</span> <span class="st">'RNA'</span>, slot <span class="op">=</span> <span class="st">'data'</span><span class="op">)</span></span></code></pre></div>
<p>Let us visualize the cell score data frame</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">my_scores</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10 × 5</span></span></span>
<span><span class="co">#&gt;    barcode          celltype   score   p_value statistic</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> AAACGAAAGAGGTCGT Enterocyte -<span style="color: #BB0000;">3.20</span> 0.001<span style="text-decoration: underline;">36</span>   ulm      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> AAACGAAAGCCTCCAG Enterocyte -<span style="color: #BB0000;">3.30</span> 0.000<span style="text-decoration: underline;">974</span>  ulm      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> AAACGAAAGTTTGGCT Enterocyte -<span style="color: #BB0000;">3.88</span> 0.000<span style="text-decoration: underline;">106</span>  ulm      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> AAACGAACACAAGCTT Enterocyte -<span style="color: #BB0000;">4.31</span> 0.000<span style="text-decoration: underline;">016</span>4 ulm      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> AAACGAAGTCGCTTGG Enterocyte -<span style="color: #BB0000;">1.22</span> 0.221     ulm      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> AAACGAAGTCTTACTT Enterocyte -<span style="color: #BB0000;">4.01</span> 0.000<span style="text-decoration: underline;">062</span>2 ulm      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> AAACGAAGTTACCCTC Enterocyte -<span style="color: #BB0000;">3.00</span> 0.002<span style="text-decoration: underline;">68</span>   ulm      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> AAACGAATCACCGGTG Enterocyte -<span style="color: #BB0000;">4.01</span> 0.000<span style="text-decoration: underline;">060</span>1 ulm      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> AAACGCTAGAGTCAAT Enterocyte -<span style="color: #BB0000;">1.93</span> 0.053<span style="text-decoration: underline;">5</span>    ulm      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> AAACGCTAGATGAATC Enterocyte -<span style="color: #BB0000;">3.71</span> 0.000<span style="text-decoration: underline;">212</span>  ulm</span></span></code></pre></div>
<p>As seen, the cell score data frame contains the scores of barcodes
for each cell signature as well as the p values. Positive scores infer
enriched signatures while negative scores infer non-enriched signatures
for each barcode.</p>
<p>Next is to label each barcode by cell types based on the signature
scores. Cells with one significant signature score will have a
single-cell label while cells with more than one significant signature
score will have a multi-cell label. Please note that barcodes and cells
are used interchangeably.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_ass</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetCellAssignments.html">GetCellAssignments</a></span><span class="op">(</span>score_data <span class="op">=</span> <span class="va">my_scores</span>, cut_off <span class="op">=</span> <span class="fl">1</span>, p_val <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p>We choose a cut-off of 1 to depict enriched signatures. This means
that we considered all negative scores and those below 1 as
non-enriched. Ideally, all positive scores should be considered enriched
but choosing a cut-off of 1 will increase precision of our cell
assignment. However, users may choose to set the cut-off at 0 or higher.
Note that there is a trade off between cutoff and cell dropout.</p>
<p>Let us visualize the cell assignment data frame.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">my_ass</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;             barcode statistic count_ulm                        celltype_ulm</span></span>
<span><span class="co">#&gt; 1  AAAGGTATCGGCTTCT       ulm         2        Enterocyte_Progenitor late-1</span></span>
<span><span class="co">#&gt; 2  AACAGGGAGAAGTCAT       ulm         2        Enterocyte_Progenitor late-1</span></span>
<span><span class="co">#&gt; 3  AACCAACCATCAGCAT       ulm         2        Enterocyte_Progenitor late-1</span></span>
<span><span class="co">#&gt; 4  AACCTGAGTACCGTCG       ulm         3 Enterocyte_Paneth_Progenitor late-1</span></span>
<span><span class="co">#&gt; 5  AACGAAACAAGCGCAA       ulm         2        Enterocyte_Progenitor late-1</span></span>
<span><span class="co">#&gt; 6  AACGAAACAGGTCCCA       ulm         2        Enterocyte_Progenitor late-1</span></span>
<span><span class="co">#&gt; 7  AACGGGAAGTGACACG       ulm         2        Enterocyte_Progenitor late-1</span></span>
<span><span class="co">#&gt; 8  AACGGGATCCATACTT       ulm         2        Enterocyte_Progenitor late-1</span></span>
<span><span class="co">#&gt; 9  AACGGGATCGTAGCTA       ulm         2        Enterocyte_Progenitor late-1</span></span>
<span><span class="co">#&gt; 10 AAGAACAAGTTCTACG       ulm         2        Enterocyte_Progenitor late-1</span></span>
<span><span class="co">#&gt;      avg_pvalue avg_score</span></span>
<span><span class="co">#&gt; 1  2.523098e-13  8.319202</span></span>
<span><span class="co">#&gt; 2  2.424298e-15  8.761571</span></span>
<span><span class="co">#&gt; 3  9.232242e-16  8.127915</span></span>
<span><span class="co">#&gt; 4  8.578408e-06  7.990285</span></span>
<span><span class="co">#&gt; 5  2.859176e-09  5.994553</span></span>
<span><span class="co">#&gt; 6  1.859488e-17  8.907267</span></span>
<span><span class="co">#&gt; 7  2.196144e-06  6.576429</span></span>
<span><span class="co">#&gt; 8  5.995422e-15  8.042991</span></span>
<span><span class="co">#&gt; 9  1.101760e-17 10.222142</span></span>
<span><span class="co">#&gt; 10 2.110035e-04  4.999832</span></span></code></pre></div>
<p>As we can see, the cell assignment data frame has a dedicated column
called “count_ulm” which tells how many cell signatures a given barcode
has. For example, a count_ulm value of 2 means that the barcode has a
significant score for two different signatures and as such it is a
predicted doublet, while a value of 3 represents a predicted triplet.
Also, the “celltype_ulm” column contains the specific signatures that a
barcode expresses. For example, the first 3 barcodes have count_ulm
values of 2 and the associated cell labels are “Enterocyte_Progenitor
late-1” as seen in the celltype_ulm column. This means that those
barcodes are cells (doublets) that were enriched in dual gene
signatures.</p>
<p>Now that we have predicted each barcode in the data and assigned
labels to them based on cell type-specific gene sigatures, we can then
add the new labels and the associated statistics to the metadata of our
Seurat Object. Let us view the metadata once again before we add the
predicted labels</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">int_singData</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#&gt;                     orig.ident nCount_RNA nFeature_RNA          Cell_Type</span></span>
<span><span class="co">#&gt; AAACGAAAGAGGTCGT SeuratProject       6540         2190   Progenitor early</span></span>
<span><span class="co">#&gt; AAACGAAAGCCTCCAG SeuratProject      11193         3062  Progenitor late-1</span></span>
<span><span class="co">#&gt; AAACGAAAGTTTGGCT SeuratProject      12139         3234   Progenitor early</span></span>
<span><span class="co">#&gt; AAACGAACACAAGCTT SeuratProject      12188         3445 Transit amplifying</span></span>
<span><span class="co">#&gt; AAACGAAGTCGCTTGG SeuratProject       2493         1112  Progenitor late-2</span></span>
<span><span class="co">#&gt; AAACGAAGTCTTACTT SeuratProject       7547         2686             Goblet</span></span>
<span><span class="co">#&gt;                                       Source Subject_strain          Tissue</span></span>
<span><span class="co">#&gt; AAACGAAAGAGGTCGT Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAAAGCCTCCAG Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAAAGTTTGGCT Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAACACAAGCTT Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAAGTCGCTTGG Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAAGTCTTACTT Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt;                  Pass_QC RNA_snn_res.0.3 seurat_clusters</span></span>
<span><span class="co">#&gt; AAACGAAAGAGGTCGT    TRUE               2               2</span></span>
<span><span class="co">#&gt; AAACGAAAGCCTCCAG    TRUE               3               3</span></span>
<span><span class="co">#&gt; AAACGAAAGTTTGGCT    TRUE               0               0</span></span>
<span><span class="co">#&gt; AAACGAACACAAGCTT    TRUE               0               0</span></span>
<span><span class="co">#&gt; AAACGAAGTCGCTTGG    TRUE               6               6</span></span>
<span><span class="co">#&gt; AAACGAAGTCTTACTT    TRUE               1               1</span></span></code></pre></div>
<p>Now let us add the predicted labels and view the metadata.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">int_singData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddMetaObject.html">AddMetaObject</a></span><span class="op">(</span><span class="va">int_singData</span>, cell_class_df <span class="op">=</span> <span class="va">my_ass</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">int_singData</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#&gt;                     orig.ident nCount_RNA nFeature_RNA          Cell_Type</span></span>
<span><span class="co">#&gt; AAACGAAAGAGGTCGT SeuratProject       6540         2190   Progenitor early</span></span>
<span><span class="co">#&gt; AAACGAAAGCCTCCAG SeuratProject      11193         3062  Progenitor late-1</span></span>
<span><span class="co">#&gt; AAACGAAAGTTTGGCT SeuratProject      12139         3234   Progenitor early</span></span>
<span><span class="co">#&gt; AAACGAACACAAGCTT SeuratProject      12188         3445 Transit amplifying</span></span>
<span><span class="co">#&gt; AAACGAAGTCGCTTGG SeuratProject       2493         1112  Progenitor late-2</span></span>
<span><span class="co">#&gt; AAACGAAGTCTTACTT SeuratProject       7547         2686             Goblet</span></span>
<span><span class="co">#&gt;                                       Source Subject_strain          Tissue</span></span>
<span><span class="co">#&gt; AAACGAAAGAGGTCGT Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAAAGCCTCCAG Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAAAGTTTGGCT Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAACACAAGCTT Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAAGTCGCTTGG Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt; AAACGAAGTCTTACTT Small Intestinal Epithelium       C57BL/6J Small Intestine</span></span>
<span><span class="co">#&gt;                  Pass_QC RNA_snn_res.0.3 seurat_clusters statistic count_ulm</span></span>
<span><span class="co">#&gt; AAACGAAAGAGGTCGT    TRUE               2               2      &lt;NA&gt;        NA</span></span>
<span><span class="co">#&gt; AAACGAAAGCCTCCAG    TRUE               3               3       ulm         1</span></span>
<span><span class="co">#&gt; AAACGAAAGTTTGGCT    TRUE               0               0      &lt;NA&gt;        NA</span></span>
<span><span class="co">#&gt; AAACGAACACAAGCTT    TRUE               0               0      &lt;NA&gt;        NA</span></span>
<span><span class="co">#&gt; AAACGAAGTCGCTTGG    TRUE               6               6       ulm         2</span></span>
<span><span class="co">#&gt; AAACGAAGTCTTACTT    TRUE               1               1       ulm         1</span></span>
<span><span class="co">#&gt;                                         celltype_ulm   avg_pvalue avg_score</span></span>
<span><span class="co">#&gt; AAACGAAAGAGGTCGT                                &lt;NA&gt;           NA        NA</span></span>
<span><span class="co">#&gt; AAACGAAAGCCTCCAG                   Progenitor late-1 1.785378e-04  3.748499</span></span>
<span><span class="co">#&gt; AAACGAAAGTTTGGCT                                &lt;NA&gt;           NA        NA</span></span>
<span><span class="co">#&gt; AAACGAACACAAGCTT                                &lt;NA&gt;           NA        NA</span></span>
<span><span class="co">#&gt; AAACGAAGTCGCTTGG Progenitor late-1_Progenitor late-2 5.336437e-04  4.889417</span></span>
<span><span class="co">#&gt; AAACGAAGTCTTACTT                              Goblet 6.233197e-06  4.519871</span></span></code></pre></div>
<p>Obviously, the statistics, count_ulm, celltype_ulm, avg_pvalue and
avg_score columns have now been added to our Seurat Object. The
count_ulm and celltype_ulm columns have been explained in the previous
section. The avg_score contains the average of the predicted scores for
all signatures enriched in a barcode, while the avg_pvalue takes the
average of the p values associated with those scores. Since each p value
was statistically significant, taking the average will ensure that the
final p value remains statistically significant. The average scores and
p values are only applicable to multiplets (count_ulm &gt; 2) having
multi signature scores, singlets (count_ulm = 1) will retain their
single scores and p values. Finally, some barcodes have NAs for the
newly added column, this represent barcodes that did not significantly
enrich in any signature.</p>
<p>The final steps in the pipeline involve isolating multiplets and
plotting physical cell-cell interaction network. Let us isolate the
multiplets from our Seurat Object, setting minCells = 2, meaning we want
barcodes of doublets and above. Set minCells to 3 if you want triplets
and above etc.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_mult</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetMultiplet.html">GetMultiplet</a></span><span class="op">(</span><span class="va">int_singData</span>, minCells <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removing 1929 cells missing data for vars requested</span></span></code></pre></div>
<p>This generates a list containing a multiplet Seurat Object and a
multiplet. The multiplet summary shows multiplet types and frequency.
Let us show the multiplet summary table.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multSummary</span> <span class="op">&lt;-</span> <span class="va">my_mult</span><span class="op">$</span><span class="va">multSummary</span></span>
<span><span class="va">multSummary</span></span>
<span><span class="co">#&gt;                                            multipletType frequency</span></span>
<span><span class="co">#&gt; 1           Enterocyte_Enteroendocrine_Progenitor late-1         1</span></span>
<span><span class="co">#&gt; 2                    Enterocyte_Goblet_Progenitor late-1         3</span></span>
<span><span class="co">#&gt; 3                    Enterocyte_Paneth_Progenitor late-1         2</span></span>
<span><span class="co">#&gt; 4                           Enterocyte_Progenitor late-1       216</span></span>
<span><span class="co">#&gt; 5         Enterocyte_Progenitor late-1_Progenitor late-2        12</span></span>
<span><span class="co">#&gt; 6                                 Enteroendocrine_Goblet         3</span></span>
<span><span class="co">#&gt; 7                       Enteroendocrine_Progenitor early         1</span></span>
<span><span class="co">#&gt; 8                                          Goblet_Paneth       104</span></span>
<span><span class="co">#&gt; 9                       Goblet_Paneth_Transit amplifying         2</span></span>
<span><span class="co">#&gt; 10                                    Goblet_Paneth_Tuft         1</span></span>
<span><span class="co">#&gt; 11                               Goblet_Progenitor early        14</span></span>
<span><span class="co">#&gt; 12            Goblet_Progenitor early_Transit amplifying         1</span></span>
<span><span class="co">#&gt; 13                              Goblet_Progenitor late-1         4</span></span>
<span><span class="co">#&gt; 14                              Goblet_Progenitor late-2         1</span></span>
<span><span class="co">#&gt; 15                             Goblet_Transit amplifying        21</span></span>
<span><span class="co">#&gt; 16                                           Goblet_Tuft         3</span></span>
<span><span class="co">#&gt; 17                               Paneth_Progenitor early         3</span></span>
<span><span class="co">#&gt; 18                          Paneth_Progenitor early_Stem         1</span></span>
<span><span class="co">#&gt; 19                              Paneth_Progenitor late-1         2</span></span>
<span><span class="co">#&gt; 20                             Paneth_Transit amplifying         2</span></span>
<span><span class="co">#&gt; 21                    Progenitor early_Progenitor late-1         9</span></span>
<span><span class="co">#&gt; 22 Progenitor early_Progenitor late-1_Transit amplifying         1</span></span>
<span><span class="co">#&gt; 23                    Progenitor early_Progenitor late-2         1</span></span>
<span><span class="co">#&gt; 24                                 Progenitor early_Stem         4</span></span>
<span><span class="co">#&gt; 25              Progenitor early_Stem_Transit amplifying         1</span></span>
<span><span class="co">#&gt; 26                   Progenitor early_Transit amplifying       148</span></span>
<span><span class="co">#&gt; 27                   Progenitor late-1_Progenitor late-2       136</span></span>
<span><span class="co">#&gt; 28                  Progenitor late-1_Transit amplifying         1</span></span></code></pre></div>
<p>We can also see the total number of predicted multiplets from the
summary table</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">multSummary</span><span class="op">$</span><span class="va">frequency</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 698</span></span></code></pre></div>
<p>The multiplet Seurat Object is just a typical object that can be
utilised for multiplet modeling or any downstream analysis as the user
pleases. Let us see the multiplet Seurat Object.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multObj</span> <span class="op">&lt;-</span> <span class="va">my_mult</span><span class="op">$</span><span class="va">multObj</span></span>
<span><span class="va">multObj</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 15610 features across 698 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (15610 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span></code></pre></div>
<p>We can also see the multiplet order from the multiplet object</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">multObj</span><span class="op">$</span><span class="va">count_ulm</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   2   3 </span></span>
<span><span class="co">#&gt; 673  25</span></span></code></pre></div>
<p>There are 698 multiplets i total, of these, 673 were doublets and 25
are triplets.</p>
<p>As we can observe in the multiplet summary table, some multiplet
types (e.g Enterocyte_Paneth_Progenitor.late.1) have a frequency of 1,
meaning only one cell (barcode) belong to this multiplet type, occuring
only once in the dataset. However, some multiplet types have a frequency
well over 100, meaning that these types are found in well over 100
barcodes. To improve confidence, it is imperative to filter the obtained
multiplets to discard those that are not found in sufficient number of
barcodes. Now let us filter our multiplets to include only those
muliplet types that were found in at least 10 barcodes (minFreq =
10).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_mult_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterMultiplet.html">FilterMultiplet</a></span><span class="op">(</span><span class="va">int_singData</span>, minCells <span class="op">=</span> <span class="fl">2</span>, minFreq <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removing 1929 cells missing data for vars requested</span></span></code></pre></div>
<p>Similarly, this returns a list of filtered multiplet summary table
and filtered multiplet Seurat Object. The filtered table will now
contain only multiplet types with frequency of 10 and above. Also the
multiplet types have now reduced from 28 to 6 after filtering.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multSummaryFilt</span> <span class="op">&lt;-</span> <span class="va">my_mult_filt</span><span class="op">$</span><span class="va">multSummaryFilt</span></span>
<span><span class="va">multSummaryFilt</span></span>
<span><span class="co">#&gt;                                     multipletType frequency</span></span>
<span><span class="co">#&gt; 4                    Enterocyte_Progenitor late-1       216</span></span>
<span><span class="co">#&gt; 5  Enterocyte_Progenitor late-1_Progenitor late-2        12</span></span>
<span><span class="co">#&gt; 8                                   Goblet_Paneth       104</span></span>
<span><span class="co">#&gt; 11                        Goblet_Progenitor early        14</span></span>
<span><span class="co">#&gt; 15                      Goblet_Transit amplifying        21</span></span>
<span><span class="co">#&gt; 26            Progenitor early_Transit amplifying       148</span></span>
<span><span class="co">#&gt; 27            Progenitor late-1_Progenitor late-2       136</span></span></code></pre></div>
<p>We can similarly explore the filtered multiplet object. Again, this
can be used for downstream analysis as a typical scRNAseq data if the
user pleases.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multObjFilt</span> <span class="op">&lt;-</span> <span class="va">my_mult_filt</span><span class="op">$</span><span class="va">multObjFilt</span></span>
<span><span class="va">multObjFilt</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 15610 features across 651 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (15610 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span></code></pre></div>
<p>Finally, we will plot a cell-cell interaction network from the
filtered multiplets. To do this, we need to first decompose the filtered
multiplet summary table in to a pairwise node-edge data frame.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_node_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetNodeDF.html">GetNodeDF</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="va">multSummaryFilt</span><span class="op">)</span></span></code></pre></div>
<p>Let us view the pairwise node-edge dataframeformed</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_node_df</span></span>
<span><span class="co">#&gt;               Cell1              Cell2 n_cells</span></span>
<span><span class="co">#&gt; 1            Goblet             Paneth     104</span></span>
<span><span class="co">#&gt; 2            Goblet   Progenitor early      14</span></span>
<span><span class="co">#&gt; 3        Enterocyte  Progenitor late-1     228</span></span>
<span><span class="co">#&gt; 4        Enterocyte  Progenitor late-2      12</span></span>
<span><span class="co">#&gt; 5 Progenitor late-1  Progenitor late-2     148</span></span>
<span><span class="co">#&gt; 6            Goblet Transit amplifying      21</span></span>
<span><span class="co">#&gt; 7  Progenitor early Transit amplifying     148</span></span></code></pre></div>
<p>Finally, we can plot the physical interaction network</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotNetwork.html">PlotNetwork</a></span><span class="op">(</span><span class="va">my_node_df</span><span class="op">)</span></span></code></pre></div>
<p><img src="ULM-vignette_files/figure-html/unnamed-chunk-24-1.png" width="768" style="display: block; margin: auto auto auto 0;"></p>
<p>Indeed, we see interactions involving progenitors and other
intestinal epithelial cells. This is biologically plausible since
progenitor cells are known to transition through stages to replace the
rapidly old intestinal epithelial cells such as enterocyte, transit
amplifying cells and goblet cells.</p>
<p>In the final part of this vignette, we show that it is also possible
to use a reference scRNAseq data set to generate signatures which can
then be utilised to predict multiplet and depict physical interactions
in a different query scRNAseq data of similar tissue of origin. For this
purpose, we use the intestinal scRNAseq data above as reference to
generate signatures. We will then utilise this to predict multiplets in
a query dataset of partially dissociated intestinal scRNAseq data from
the same study (Andrews et al., 2021, pubmed: GSE175664). We load query
and reference data sets</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"int_singData"</span><span class="op">)</span></span>
<span><span class="va">ref_obj</span> <span class="op">&lt;-</span> <span class="va">int_singData</span></span>
<span><span class="va">ref_obj</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 15610 features across 5279 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (15610 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"int_multData"</span><span class="op">)</span></span>
<span><span class="va">query_obj</span> <span class="op">&lt;-</span> <span class="va">int_multData</span></span>
<span><span class="va">query_obj</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 15615 features across 3671 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (15615 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span></code></pre></div>
<p>Let us generate signatures from the reference dataset</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">101324</span><span class="op">)</span></span>
<span><span class="va">ref_sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetSignature.html">GetSignature</a></span><span class="op">(</span><span class="va">ref_obj</span>, ident_col <span class="op">=</span> <span class="va">ref_obj</span><span class="op">$</span><span class="va">Cell_Type</span><span class="op">)</span></span>
<span><span class="co">#&gt; using the specified seurat ident to generate signatures</span></span>
<span><span class="co">#&gt; Calculating cluster Progenitor early</span></span>
<span><span class="co">#&gt; Calculating cluster Progenitor late-1</span></span>
<span><span class="co">#&gt; Calculating cluster Transit amplifying</span></span>
<span><span class="co">#&gt; Calculating cluster Progenitor late-2</span></span>
<span><span class="co">#&gt; Calculating cluster Goblet</span></span>
<span><span class="co">#&gt; Calculating cluster Stem</span></span>
<span><span class="co">#&gt; Calculating cluster Enterocyte</span></span>
<span><span class="co">#&gt; Calculating cluster Paneth</span></span>
<span><span class="co">#&gt; Calculating cluster Enteroendocrine</span></span>
<span><span class="co">#&gt; Calculating cluster Tuft</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ref_sig</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10 × 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   source [1]</span></span></span>
<span><span class="co">#&gt;    source           target          mor</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Progenitor early Ccna1             1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Progenitor early Hist1h2br         1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Progenitor early BC051077          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Progenitor early Gm10501           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Progenitor early C330021F23Rik     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Progenitor early Cdkn3             1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Progenitor early Cdc25c            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Progenitor early Ccnb2             1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Progenitor early Cdc20             1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Progenitor early Lipg              1</span></span></code></pre></div>
<p>NB: It is also possible to directly import and utilize a reference
signature but this must be formatted to suit the output of the
GetSignature() function above. That is, it must be a data frame (or
tibble) with 3 columns- source, target and mor, containing the cell
type, signatures and weight (=1) respectively.</p>
<p>We then utilize the reference signature to predict multiplets in the
query data set and construct a physical cell-cell interaction
network.</p>
<p>Getting query cell scores</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetCellScores.html">GetCellScores</a></span><span class="op">(</span>seurat_obj <span class="op">=</span> <span class="va">query_obj</span>, signatures <span class="op">=</span> <span class="va">ref_sig</span>, assay <span class="op">=</span> <span class="st">'RNA'</span>, slot <span class="op">=</span> <span class="st">'data'</span><span class="op">)</span></span></code></pre></div>
<p>Assigning cell labels</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_ass</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetCellAssignments.html">GetCellAssignments</a></span><span class="op">(</span>score_data <span class="op">=</span> <span class="va">my_scores</span><span class="op">)</span></span></code></pre></div>
<p>Adding cell labels to the metadata</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lab_query_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddMetaObject.html">AddMetaObject</a></span><span class="op">(</span><span class="va">query_obj</span>, cell_class_df <span class="op">=</span> <span class="va">my_ass</span><span class="op">)</span></span></code></pre></div>
<p>Obtaining predicted multiplets</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_mult</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetMultiplet.html">GetMultiplet</a></span><span class="op">(</span><span class="va">lab_query_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removing 830 cells missing data for vars requested</span></span></code></pre></div>
<p>Multiplet summary table</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_multSummary</span> <span class="op">&lt;-</span> <span class="va">query_mult</span><span class="op">$</span><span class="va">multSummary</span></span>
<span><span class="va">query_multSummary</span></span>
<span><span class="co">#&gt;                                            multipletType frequency</span></span>
<span><span class="co">#&gt; 1           Enterocyte_Enteroendocrine_Progenitor late-1         1</span></span>
<span><span class="co">#&gt; 2                    Enterocyte_Goblet_Progenitor late-1         9</span></span>
<span><span class="co">#&gt; 3  Enterocyte_Goblet_Progenitor late-1_Progenitor late-2         1</span></span>
<span><span class="co">#&gt; 4                    Enterocyte_Paneth_Progenitor late-1         5</span></span>
<span><span class="co">#&gt; 5  Enterocyte_Paneth_Progenitor late-1_Progenitor late-2         1</span></span>
<span><span class="co">#&gt; 6                           Enterocyte_Progenitor late-1       256</span></span>
<span><span class="co">#&gt; 7         Enterocyte_Progenitor late-1_Progenitor late-2        57</span></span>
<span><span class="co">#&gt; 8                      Enterocyte_Progenitor late-1_Tuft         2</span></span>
<span><span class="co">#&gt; 9                                 Enteroendocrine_Goblet         1</span></span>
<span><span class="co">#&gt; 10                         Enteroendocrine_Goblet_Paneth         1</span></span>
<span><span class="co">#&gt; 11                      Enteroendocrine_Progenitor early         6</span></span>
<span><span class="co">#&gt; 12                     Enteroendocrine_Progenitor late-1         1</span></span>
<span><span class="co">#&gt; 13                    Enteroendocrine_Transit amplifying         2</span></span>
<span><span class="co">#&gt; 14                                         Goblet_Paneth       211</span></span>
<span><span class="co">#&gt; 15                        Goblet_Paneth_Progenitor early         2</span></span>
<span><span class="co">#&gt; 16                               Goblet_Progenitor early        83</span></span>
<span><span class="co">#&gt; 17            Goblet_Progenitor early_Transit amplifying         3</span></span>
<span><span class="co">#&gt; 18                              Goblet_Progenitor late-1        18</span></span>
<span><span class="co">#&gt; 19            Goblet_Progenitor late-1_Progenitor late-2         5</span></span>
<span><span class="co">#&gt; 20                              Goblet_Progenitor late-2         8</span></span>
<span><span class="co">#&gt; 21                             Goblet_Transit amplifying        25</span></span>
<span><span class="co">#&gt; 22                               Paneth_Progenitor early        15</span></span>
<span><span class="co">#&gt; 23            Paneth_Progenitor early_Transit amplifying         2</span></span>
<span><span class="co">#&gt; 24                              Paneth_Progenitor late-1         2</span></span>
<span><span class="co">#&gt; 25            Paneth_Progenitor late-1_Progenitor late-2         4</span></span>
<span><span class="co">#&gt; 26                                           Paneth_Stem         4</span></span>
<span><span class="co">#&gt; 27                             Paneth_Transit amplifying         9</span></span>
<span><span class="co">#&gt; 28                    Progenitor early_Progenitor late-1         6</span></span>
<span><span class="co">#&gt; 29                    Progenitor early_Progenitor late-2         1</span></span>
<span><span class="co">#&gt; 30                   Progenitor early_Transit amplifying        79</span></span>
<span><span class="co">#&gt; 31                                 Progenitor early_Tuft         3</span></span>
<span><span class="co">#&gt; 32                   Progenitor late-1_Progenitor late-2       174</span></span>
<span><span class="co">#&gt; 33              Progenitor late-1_Progenitor late-2_Tuft         2</span></span></code></pre></div>
<p>Multiplet order</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_multObj</span> <span class="op">&lt;-</span> <span class="va">query_mult</span><span class="op">$</span><span class="va">multObj</span></span>
<span><span class="va">query_multObj</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 15615 features across 999 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (15615 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">query_multObj</span><span class="op">$</span><span class="va">count_ulm</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   2   3   4 </span></span>
<span><span class="co">#&gt; 904  93   2</span></span></code></pre></div>
<p>Filtering multiplets</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_mult_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterMultiplet.html">FilterMultiplet</a></span><span class="op">(</span><span class="va">lab_query_obj</span>, minFreq <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removing 830 cells missing data for vars requested</span></span></code></pre></div>
<p>Filtered multiplet summary table</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_multSummaryFilt</span> <span class="op">&lt;-</span> <span class="va">query_mult_filt</span><span class="op">$</span><span class="va">multSummaryFilt</span></span>
<span><span class="va">query_multSummaryFilt</span></span>
<span><span class="co">#&gt;                                     multipletType frequency</span></span>
<span><span class="co">#&gt; 2             Enterocyte_Goblet_Progenitor late-1         9</span></span>
<span><span class="co">#&gt; 6                    Enterocyte_Progenitor late-1       256</span></span>
<span><span class="co">#&gt; 7  Enterocyte_Progenitor late-1_Progenitor late-2        57</span></span>
<span><span class="co">#&gt; 14                                  Goblet_Paneth       211</span></span>
<span><span class="co">#&gt; 16                        Goblet_Progenitor early        83</span></span>
<span><span class="co">#&gt; 18                       Goblet_Progenitor late-1        18</span></span>
<span><span class="co">#&gt; 20                       Goblet_Progenitor late-2         8</span></span>
<span><span class="co">#&gt; 21                      Goblet_Transit amplifying        25</span></span>
<span><span class="co">#&gt; 22                        Paneth_Progenitor early        15</span></span>
<span><span class="co">#&gt; 27                      Paneth_Transit amplifying         9</span></span>
<span><span class="co">#&gt; 30            Progenitor early_Transit amplifying        79</span></span>
<span><span class="co">#&gt; 32            Progenitor late-1_Progenitor late-2       174</span></span></code></pre></div>
<p>Generating pairwise node-edge data frame from the filtered multiplet
summary table</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query_network_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetNodeDF.html">GetNodeDF</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="va">query_multSummaryFilt</span><span class="op">)</span></span></code></pre></div>
<p>Plotting query cell-cell interaction network</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotNetwork.html">PlotNetwork</a></span><span class="op">(</span><span class="va">query_network_df</span><span class="op">)</span></span></code></pre></div>
<p><img src="ULM-vignette_files/figure-html/unnamed-chunk-38-1.png" width="1440" style="display: block; margin: auto auto auto 0;"></p>
<p>THE END</p>
<p>Reference</p>
<p>Andrews N, Serviss JT, Geyer N, Andersson AB, Dzwonkowska E, Šutevski
I, et al. An unsupervised method for physical cell interaction profiling
of complex tissues. Nat Methods. 2021;18:912–20.</p>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Sodiq A. Hameed, Luis F. Iglesias-Martinez, Walter Kolch, Vadim Zhernovkov.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
